nibabel/cifti2/cifti2.py

```diff
@@ -0,0 +1,1095 @@
+        for p in self:
+            if isinstance(p, Cifti2Volume):
+                return p
+        return None
+
+    @volume.setter
+    def volume(self, volume):
+        if not isinstance(volume, Cifti2Volume):
+            raise ValueError("You can only set a volume with a volume")
+        for i, v in enumerate(self):
+            if isinstance(v, Cifti2Volume):
+                break
+        else:
+            self.append(volume)
+            return
+        self[i] = volume
+
+    @volume.deleter
+    def volume(self):
+        for i, v in enumerate(self):
+            if isinstance(v, Cifti2Volume):
+                break
+        else:
+            raise ValueError("No Cifti2Volume element")
+        del self[i]
+
+    @property
+    def brain_models(self):
+        for p in self:
+            if isinstance(p, Cifti2BrainModel):
+                yield p
+
+    def _to_xml_element(self):
+        if self.applies_to_matrix_dimension is None:
+            raise CIFTI2HeaderError(
+                'MatrixIndicesMap element requires to be applied to at least 1 dimension'
+            )
+
+        mat_ind_map = xml.Element('MatrixIndicesMap')
+        dims_as_strings = [str(dim) for dim in self.applies_to_matrix_dimension]
+        mat_ind_map.attrib['AppliesToMatrixDimension'] = ','.join(dims_as_strings)
+        for key in ['IndicesMapToDataType', 'NumberOfSeriesPoints', 'SeriesExponent',
+                    'SeriesStart', 'SeriesStep', 'SeriesUnit']:
+            attr = _underscore(key)
+            value = getattr(self, attr)
+            if value is not None:
+                mat_ind_map.attrib[key] = str(value)
+        for map_ in self:
+            mat_ind_map.append(map_._to_xml_element())
+
+        return mat_ind_map
+
+
+class Cifti2Matrix(xml.XmlSerializable, collections.MutableSequence):
+    def __init__(self):
+        self._mims = []
+        self.metadata = None
+
+    @property
+    def metadata(self):
+        return self._meta
+
+    @metadata.setter
+    def metadata(self, meta):
+        """ Set the metadata for this Cifti2Header
+
+        Parameters
+        ----------
+        meta : Cifti2MetaData
+
+        Returns
+        -------
+        None
+        """
+        self._meta = _value_if_klass(meta, Cifti2MetaData)
+
+    def __setitem__(self, key, value):
+        if not isinstance(value, Cifti2MatrixIndicesMap):
+            raise TypeError("Not a valid Cifti2MatrixIndicesMap instance")
+        self._mims[key] = value
+
+    def __getitem__(self, key):
+        return self._mims[key]
+
+    def __delitem__(self, key):
+        del self._mims[key]
+
+    def __len__(self):
+        return len(self._mims)
+
+    def insert(self, index, value):
+        if not isinstance(value, Cifti2MatrixIndicesMap):
+            raise TypeError("Not a valid Cifti2MatrixIndicesMap instance")
+        self._mims.insert(index, value)
+
+    def _to_xml_element(self):
+        if (len(self) == 0 and self.metadata is None):
```

Comment:
Yes, I don't want to slow down the merge.  Some example code using the API would make it much clearer to me how it works (I am not familiar with python), and suggest ways to improve the API (and can double as additional tests).

As a start, though, from experience with a serialized-form API on CIFTI, I can immediately suggest two added helper member functions onto the XML object (prototypes in pseudo-c++) :

`getMapping(int dimension)` - searches through the mapping elements, and their AppliesToMatrixDimension list, and returns the mapping that matches the argument (otherwise throws?)

`setMapping(int dimension, Mapping input)` - searches similarly, removes the matching AppliesToMatrixDimension element, deletes the mapping if that emptied the AppliesToMatrixDimension list, then appends the input mapping to the MatrixIndicesMap list, and assigns the dimension to its AppliesToMatrixDimension list